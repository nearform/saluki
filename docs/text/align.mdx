import { Align } from './'

# Text Align

---

## Example

<Align>Hello</Align>

```js
const Text = styled.h1`
  ${text.align.center}
`
```

## Rules

```js
text.align.center
text.align.left
text.align.right
text.align.justify
```
