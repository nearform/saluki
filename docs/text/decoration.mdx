import { Decoration } from './'

# Text Decoration

---

## Example

<Decoration>Hello</Decoration>

```jsx
const Text = styled.h1`
  ${text.decoration.underline}
`
```

## Configuration

Configurable via;

```
text.decoration
```
