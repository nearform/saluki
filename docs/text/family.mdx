import { Family } from './'

# Text Family

---

## Example

<Family>Hello</Family>

```jsx
const Text = styled.h1`
  ${text.serif}
`
```

## Rules

```js
text.sans
text.serif
```

## Configuration

Configurable via;

```
text
```

## Defaults

```js
{
  text: {
    sans: 'sans',
    serif: 'serif'
  }
}
```
