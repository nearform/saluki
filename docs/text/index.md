# Text

---

## Configuration

Configurable via;

```
text
```

## Defaults

```js
{
  size: {
    small: '1rem',
    medium: '2rem',
    large: '3rem'
  },
  weight: {
    thin: '200',
    normal: 'normal',
    bold: 'bold'
  },
  align: {
    center: 'center',
    left: 'left',
    right: 'right',
    justify: 'justify'
  },
  transform: {
    uppercase: 'uppercase',
    lowercase: 'lowercase',
    capitalize: 'capitalize',
    none: 'none'
  },
  style: {
    normal: 'normal',
    italic: 'italic',
    oblique: 'oblique'
  },
  decoration: {
    underline: 'underline',
    lineThrough: 'line-through',
    none: 'none'
  },
  lineHeight: {
    small: 1,
    medium: 1.5,
    large: 2
  },
  letterSpacing: {
    small: '.075em',
    medium: '.125em',
    large: '.2em'
  },
  sans: 'sans',
  serif: 'serif'
}
```
