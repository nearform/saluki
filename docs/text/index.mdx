# Text

---

## Rules

```
text.family.sans
text.family.serif

text.size.small
text.size.medium
text.size.large

text.weight.thin
text.weight.normal
text.weight.bold

text.align.center
text.align.left
text.align.right
text.align.justify

text.transform.uppercase
text.transform.lowercase
text.transform.capitalize
text.transform.none

text.style.italic
text.style.oblique
text.style.normal

text.decoration.underline
text.decoration.lineThrough
text.decoration.normal

text.lineHeight.none
text.lineHeight.small
text.lineHeight.medium
text.lineHeight.large

text.letterSpacing.small
text.letterSpacing.none
text.letterSpacing.medium
text.letterSpacing.large
```

## Configuration

Configurable via;

```
text
```

## Defaults

```js
{
  family: {
    sans:
      '-apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica Neue, Arial, Noto Sans, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol, Noto Color Emoji',
    serif: 'Georgia, Times, Times New Roman, serif',
  },
  size: {
    small: '1rem',
    medium: '1.25rem',
    large: '2.5rem'
  },
  weight: {
    thin: '200',
    normal: 'normal',
    bold: '700'
  },
  lineHeight: {
    none: '1',
    small: '1.25',
    medium: '1.5',
    large: '2'
  },
  letterSpacing: {
    small: '-.05em',
    none: '0',
    medium: '.025em',
    large: '.05em'
  }
}
```
