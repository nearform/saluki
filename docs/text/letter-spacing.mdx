import { LetterSpacing } from './'

# Letter Spacing

---

## Example

<LetterSpacing>Hello</LetterSpacing>

```jsx
const Text = styled.h1`
  ${text.letterSpacing.large}
`
```

## Rules

```js
text.letterSpacing.small
text.letterSpacing.medium
text.letterSpacing.large
```

## Configuration

Configurable via;

```
text.letterSpacing
```

## Defaults

```js
{
  small: '.075em',
  medium: '.125em',
  large: '.2em'
}
```
