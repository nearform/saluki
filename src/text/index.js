export const defaultRules = {
  size: {
    small: '1rem',
    medium: '2rem',
    large: '3rem'
  },
  weight: {
    thin: '200',
    normal: 'normal',
    bold: 'bold'
  },
  transform: {
    uppercase: 'uppercase',
    lowercase: 'lowercase',
    capitalize: 'capitalize',
    none: 'none'
  },
  style: {
    normal: 'normal',
    italic: 'italic',
    oblique: 'oblique'
  },
  decoration: {
    underline: 'underline',
    lineThrough: 'line-through',
    none: 'none'
  },
  lineHeight: {
    small: 1,
    medium: 1.5,
    large: 2
  },
  letterSpacing: {
    small: '.075em',
    medium: '.125em',
    large: '.2em'
  },
  sans: 'sans',
  serif: 'serif'
}

export function combineRules(
  defaultRules,
  customRules = {
    size: {},
    weight: {},
    transform: {},
    style: {},
    decoration: {},
    lineHeight: {},
    letterSpacing: {}
  }
) {
  return {
    ...defaultRules,
    ...customRules,
    size: {
      ...defaultRules.size,
      ...customRules.size
    },
    weight: {
      ...defaultRules.weight,
      ...customRules.weight
    },
    transform: {
      ...defaultRules.transform,
      ...customRules.transform
    },
    style: {
      ...defaultRules.style,
      ...customRules.style
    },
    decoration: {
      ...defaultRules.decoration,
      ...customRules.decoration
    },
    lineHeight: {
      ...defaultRules.lineHeight,
      ...customRules.lineHeight
    },
    letterSpacing: {
      ...defaultRules.letterSpacing,
      ...customRules.letterSpacing
    }
  }
}

function generateRules(rules, cssProp) {
  const newRules = {}

  for (const [key, value] of Object.entries(rules)) {
    newRules[key] = {
      [cssProp]: value
    }
  }

  return newRules
}

function generateFamilyRules(rule) {
  return {
    'font-family': rule
  }
}

export default function(rules) {
  return {
    size: generateRules(rules.size, 'font-size'),
    weight: generateRules(rules.weight, 'font-weight'),
    transform: generateRules(rules.transform, 'text-transform'),
    style: generateRules(rules.style, 'text-style'),
    decoration: generateRules(rules.decoration, 'text-decoration'),
    lineHeight: generateRules(rules.lineHeight, 'line-height'),
    letterSpacing: generateRules(rules.letterSpacing, 'letter-spacing'),
    sans: generateFamilyRules(rules.sans),
    serif: generateFamilyRules(rules.serif)
  }
}
