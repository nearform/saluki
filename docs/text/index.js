import styled from 'styled-components/macro'
import { color, text } from '../../src'

export const Color = styled.h1`
  ${color.blue}
`

export const Family = styled.h1`
  ${text.family.serif}
`

export const Size = styled.h1`
  ${text.size.large}
`

export const Weight = styled.h1`
  ${text.weight.thin}
`

export const Align = styled.h1`
  ${text.align.center}
`

export const Transform = styled.h1`
  ${text.transform.uppercase}
`

export const Style = styled.h1`
  ${text.style.italic}
`

export const Decoration = styled.h1`
  ${text.decoration.underline}
`

export const LineHeight = styled.p`
  ${text.lineHeight.large}
`

export const LetterSpacing = styled.h1`
  ${text.letterSpacing.small}
`
