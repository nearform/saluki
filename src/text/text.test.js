import text, { defaultRules, combineRules } from './index'

describe('text', () => {
  describe('when no custom config is provided', () => {
    it('should return the default config', () => {
      const combinedRules = combineRules(defaultRules)
      expect(text(combinedRules)).toMatchSnapshot()
    })
  })

  describe('when custom config is provided', () => {
    describe('when the custom config overrides a value', () => {
      it('should replace the default value', () => {
        const combinedRules = combineRules(defaultRules, {
          size: { small: '.5rem' }
        })
        expect(text(combinedRules)).toMatchSnapshot()
      })
    })

    describe('when the custom config provides a new value', () => {
      it('should add the value and not amend any existing values', () => {
        const combinedRules = combineRules(defaultRules, {
          size: { extrasmall: '.25rem' }
        })
        expect(text(combinedRules)).toMatchSnapshot()
      })
    })
  })
})
