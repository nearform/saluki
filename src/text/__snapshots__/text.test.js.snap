// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text when custom config is provided when the custom config overrides a value should replace the default value 1`] = `
Object {
  "decoration": Object {
    "lineThrough": Object {
      "text-decoration": "line-through",
    },
    "none": Object {
      "text-decoration": "none",
    },
    "underline": Object {
      "text-decoration": "underline",
    },
  },
  "letterSpacing": Object {
    "large": Object {
      "letter-spacing": ".2em",
    },
    "medium": Object {
      "letter-spacing": ".125em",
    },
    "small": Object {
      "letter-spacing": ".075em",
    },
  },
  "lineHeight": Object {
    "large": Object {
      "line-height": 2,
    },
    "medium": Object {
      "line-height": 1.5,
    },
    "small": Object {
      "line-height": 1,
    },
  },
  "sans": Object {
    "font-family": "sans",
  },
  "serif": Object {
    "font-family": "serif",
  },
  "size": Object {
    "large": Object {
      "font-size": "3rem",
    },
    "medium": Object {
      "font-size": "2rem",
    },
    "small": Object {
      "font-size": ".5rem",
    },
  },
  "style": Object {
    "italic": Object {
      "text-style": "italic",
    },
    "normal": Object {
      "text-style": "normal",
    },
    "oblique": Object {
      "text-style": "oblique",
    },
  },
  "transform": Object {
    "capitalize": Object {
      "text-transform": "capitalize",
    },
    "lowercase": Object {
      "text-transform": "lowercase",
    },
    "none": Object {
      "text-transform": "none",
    },
    "uppercase": Object {
      "text-transform": "uppercase",
    },
  },
  "weight": Object {
    "bold": Object {
      "font-weight": "bold",
    },
    "normal": Object {
      "font-weight": "normal",
    },
    "thin": Object {
      "font-weight": "200",
    },
  },
}
`;

exports[`text when custom config is provided when the custom config provides a new value should add the value and not amend any existing values 1`] = `
Object {
  "decoration": Object {
    "lineThrough": Object {
      "text-decoration": "line-through",
    },
    "none": Object {
      "text-decoration": "none",
    },
    "underline": Object {
      "text-decoration": "underline",
    },
  },
  "letterSpacing": Object {
    "large": Object {
      "letter-spacing": ".2em",
    },
    "medium": Object {
      "letter-spacing": ".125em",
    },
    "small": Object {
      "letter-spacing": ".075em",
    },
  },
  "lineHeight": Object {
    "large": Object {
      "line-height": 2,
    },
    "medium": Object {
      "line-height": 1.5,
    },
    "small": Object {
      "line-height": 1,
    },
  },
  "sans": Object {
    "font-family": "sans",
  },
  "serif": Object {
    "font-family": "serif",
  },
  "size": Object {
    "extrasmall": Object {
      "font-size": ".25rem",
    },
    "large": Object {
      "font-size": "3rem",
    },
    "medium": Object {
      "font-size": "2rem",
    },
    "small": Object {
      "font-size": "1rem",
    },
  },
  "style": Object {
    "italic": Object {
      "text-style": "italic",
    },
    "normal": Object {
      "text-style": "normal",
    },
    "oblique": Object {
      "text-style": "oblique",
    },
  },
  "transform": Object {
    "capitalize": Object {
      "text-transform": "capitalize",
    },
    "lowercase": Object {
      "text-transform": "lowercase",
    },
    "none": Object {
      "text-transform": "none",
    },
    "uppercase": Object {
      "text-transform": "uppercase",
    },
  },
  "weight": Object {
    "bold": Object {
      "font-weight": "bold",
    },
    "normal": Object {
      "font-weight": "normal",
    },
    "thin": Object {
      "font-weight": "200",
    },
  },
}
`;

exports[`text when no custom config is provided should return the default config 1`] = `
Object {
  "decoration": Object {
    "lineThrough": Object {
      "text-decoration": "line-through",
    },
    "none": Object {
      "text-decoration": "none",
    },
    "underline": Object {
      "text-decoration": "underline",
    },
  },
  "letterSpacing": Object {
    "large": Object {
      "letter-spacing": ".2em",
    },
    "medium": Object {
      "letter-spacing": ".125em",
    },
    "small": Object {
      "letter-spacing": ".075em",
    },
  },
  "lineHeight": Object {
    "large": Object {
      "line-height": 2,
    },
    "medium": Object {
      "line-height": 1.5,
    },
    "small": Object {
      "line-height": 1,
    },
  },
  "sans": Object {
    "font-family": "sans",
  },
  "serif": Object {
    "font-family": "serif",
  },
  "size": Object {
    "large": Object {
      "font-size": "3rem",
    },
    "medium": Object {
      "font-size": "2rem",
    },
    "small": Object {
      "font-size": "1rem",
    },
  },
  "style": Object {
    "italic": Object {
      "text-style": "italic",
    },
    "normal": Object {
      "text-style": "normal",
    },
    "oblique": Object {
      "text-style": "oblique",
    },
  },
  "transform": Object {
    "capitalize": Object {
      "text-transform": "capitalize",
    },
    "lowercase": Object {
      "text-transform": "lowercase",
    },
    "none": Object {
      "text-transform": "none",
    },
    "uppercase": Object {
      "text-transform": "uppercase",
    },
  },
  "weight": Object {
    "bold": Object {
      "font-weight": "bold",
    },
    "normal": Object {
      "font-weight": "normal",
    },
    "thin": Object {
      "font-weight": "200",
    },
  },
}
`;
